### PR TITLE
Fix unexpected redeploy when no changes in devfile

### PR DESCRIPTION
**Pull request by rajivnathan**
Signed-off-by: Rajiv Senthilnathan <rajivsen@ca.ibm.com>

**What type of PR is this?**
/kind bug

**What does does this PR do / why we need it**:
The ordering of the volumes was inconsistent from one push to another. This PR changes to iterate over slices instead of structs to ensure ordering is consistent. See https://github.com/openshift/odo/issues/2662 for more details.

**Which issue(s) this PR fixes**:
Fixes https://github.com/openshift/odo/issues/2662

**How to test changes / Special notes to the reviewer**:
1. Use a project with a devfile that defines at least 3 volumes
2. Run `kubectl get rs -w` in a separate terminal to watch for any ReplicaSet changes. If a redeploy occurs you will see new events during the push.
3. Run `odo push` and observe that the watch output creates some ReplicaSets
4. After the last push has completed, run `odo push` again
5. The watch output should not change

Sample devfile:
```
---
apiVersion: 1.0.0
metadata:
  generateName: openLiberty
projects:
  -
    name: openLiberty
    source:
      type: git
      location: "https://github.com/rajivnathan/openLiberty.git"
components:
  -
    type: chePlugin
    id: redhat/java/latest
    memoryLimit: 1512Mi
  -
    type: dockerimage
    image: jeevandroid/open-liberty-dev:latest
    alias: runtime
    memoryLimit: 768Mi
    command: ['tail']
    args: [ '-f', '/dev/null']
    env:
      - name: "MODE"
        value: "TEST"
      - name: "myprop"
        value: "myval"
    endpoints:
      - name: '9090/tcp'
        port: 9090
    mountSources: true
    volumes:
      - name: m2
        containerPath: /home/user/.m2
      - name: m3
        containerPath: /home/user/.m3
      - name: m4
        containerPath: /home/user/.m4
  commands:
  -
    name: Build
    actions:
      -
        type: exec
        component: runtime
        command: "TEST_ENV=true /artifacts/bin/build.sh"
        workdir: /projects/openLiberty
  -
    name: Run
    actions:
      -
        type: exec
        component: runtime
        command: "/artifacts/bin/run.sh"
        workdir: /projects/openLiberty
```

### CODE REVIEW

**Comment by johnmcollier** [comment]
@rajivnathan Can you rebase on master? The devfile parser changes went in and moved some things around which seems to have affected your changes.

**Comment by amitkrout** [comment]
@rajivnathan Failures are coming from your changes. Can you please revisit the changes.

**Comment by maysunfaisal** [comment]
I'm going to review this PR soon.

**Comment by girishramnani** [comment]
can you please add integration tests
/hold

File: pkg/kclient/volumes_test.go [hunk 1]
```diff
@@ -463,26 +487,26 @@ func TestAddPVCAndVolumeMount(t *testing.T) {
...
 						// check if container has the correct number of volume mounts
-						if len(container.VolumeMounts) != len(testContainerVolumes) {
-							t.Errorf("Incorrect number of Volume Mounts found in the pod template spec container %v, expected: %v found: %v", container.Name, len(testContainerVolumes), len(container.VolumeMounts))
+						if len(container.VolumeMounts) != len(pair.Volumes) {
+							t.Errorf("Incorrect number of Volume Mounts found in the pod template spec container %v, expected: %v found: %v", container.Name, len(pair.Volumes), len(container.VolumeMounts))
 						}
 
 						// check if container has the specified volume
 						volumeMatched := 0
 						for _, volumeMount := range container.VolumeMounts {
```
Comment (girishramnani) [inline pkg/kclient/volumes_test.go:499]: please extract this for loop into a separate function ( reducing depth )

**Comment by johnmcollier** [comment]
@rajivnathan Are you able to rebase this PR when you get a chance? Thanks!

**Comment by rajivnathan** [comment]
Spoke to @cdrage about this one. Someone from the odo team can take this over since the PR needs to be updated for devfile 2.0 changes that I'm not familiar with.

**Comment by maysunfaisal** [comment]
@rajivnathan I can take a poke at rebasing your branch. Can I have contributor privilege to your fork? Or else someone can just copy-paste the logic from this PR and create a new branch

**Comment by rajivnathan** [comment]
@maysunfaisal I spoke to @cdrage about it. I had an updated branch here: https://github.com/openshift/odo/compare/master...rajivnathan:fixRedeployBug_updated?expand=1#diff-cf36d77eeec17dc66d270cc6219b5a1cR129

I tried to update the fix but got stuck because it appears that components no longer have aliases. I'm guessing this is related to devfile 2.0 changes which I'm not really familiar with so maybe it makes sense for someone who is familiar with the devfile 2.0 changes to take over from here. They could use this new branch as a starting point to speed things up if they wanted to. It's just one commit so you could just cherry-pick the commit to your own branch. However, check with @cdrage just to make sure he or someone else hasn't already started working on it.

**Pull request closed without merging**